### PR TITLE
Add WinML Experimental API to Register ORT CustomOps Libraries

### PR DESCRIPTION
**Pull request by smk2007**
Add WinML Experimental API to Register ORT CustomOps Libraries

### CODE REVIEW

File: winml/api/Microsoft.AI.MachineLearning.Experimental.idl [hunk 1]
```diff
@@ -27,7 +27,13 @@ namespace ROOT_NS.AI.MachineLearning.Experimental {
   [marshaling_behavior(agile)]
   [dualapipartition(1)]
   runtimeclass LearningModelSessionOptionsExperimental {
+    LearningModelSessionOptionsExperimental(ROOT_NS.AI.MachineLearning.LearningModelSessionOptions options);
+
+    //! The GetNamedDimensionOverrides method retrieves the named dimension overrides configured on the session options.
     Windows.Foundation.Collections.IMapView<String, UINT32> GetNamedDimensionOverrides();
+
+    //! The RegisterCustomOpsLibrary method register an onnxruntime custom operator library into session creation.
```
Comment (PatriceVignola) [inline winml/api/Microsoft.AI.MachineLearning.Experimental.idl:35]: registers

File: winml/api/Microsoft.AI.MachineLearning.Experimental.idl [hunk 2]
```diff
@@ -139,7 +145,7 @@ namespace ROOT_NS.AI.MachineLearning.Experimental {
     //! The Save method serializes the model as an ONNX model to a specified path.
     void Save(String file_name);
 
-    //! The JoinModel fuses two models by linking  outputs from the first model, to inupts of the second. 
+    //! The JoinModel fuses two models by linking  outputs from the first model, to inupts of the second.
```
Comment (PatriceVignola) [inline winml/api/Microsoft.AI.MachineLearning.Experimental.idl:148]: double spaces after `linking`, and `inupts` -> `inputs` (preexisting)

File: winml/lib/Api.Ort/OnnxruntimeEngine.cpp [hunk 3]
```diff
@@ -428,6 +428,17 @@ HRESULT OnnxruntimeEngine::RuntimeClassInitialize(OnnxruntimeEngineFactory* engi
   return S_OK;
 }
 
+OnnxruntimeEngine::~OnnxruntimeEngine() {
+  for (auto& handle : custom_op_library_handles_) {
+    FreeLibrary((HMODULE)handle);
```
Comment (PatriceVignola) [inline winml/lib/Api.Ort/OnnxruntimeEngine.cpp:433]: prefer static_cast or reinterpret_cast here

File: winml/lib/Api.Ort/OnnxruntimeEngineBuilder.cpp [hunk 4]
```diff
@@ -53,7 +53,16 @@ STDMETHODIMP OnnxruntimeEngineBuilder::CreateEngine(_Outptr_ _winml::IEngine** o
   }
 
   if (!allow_thread_spinning_) {
-    ort_api->AddSessionConfigEntry(session_options.get(), "session.intra_op.allow_spinning", "0");
+    RETURN_HR_IF_NOT_OK_MSG(ort_api->AddSessionConfigEntry(session_options.get(), "session.intra_op.allow_spinning", "0"),
+                            ort_api);
+  }
+
+  std::vector<void*> handles;
+  for (auto& path : custom_ops_lib_paths_) {
```
Comment (PatriceVignola) [inline winml/lib/Api.Ort/OnnxruntimeEngineBuilder.cpp:61]: nit: `const auto& path`

File: winml/lib/Api/LearningModelSession.cpp [hunk 5]
```diff
@@ -136,8 +136,15 @@ void LearningModelSession::Initialize() {
 
     allow_spinning = session_options_impl->GetIntraOpThreadSpinning();
     num_intra_op_threads = session_options_impl->GetIntraOpNumThreads();
+
+    const auto& paths = session_options_impl->GetCustomOpLibraryPaths();
+    for (const auto& path : paths)
+    {
```
Comment (PatriceVignola) [inline winml/lib/Api/LearningModelSession.cpp:142]: nit: doesn't match bracket style from the rest of this file

File: winml/lib/Api/LearningModelSession.h [hunk 6]
```diff
@@ -87,7 +87,13 @@ struct LearningModelSession : LearningModelSessionT<LearningModelSession, ILearn
     return &dml_ep_lock_;
   }
 
-  static winml::LearningModelSession CreateInertSession(_winml::IEngine* engine);
+  static
+  winml::LearningModelSession
+  CreateInertSession(
+    _winml::IEngine* engine);
```
Comment (PatriceVignola) [inline winml/lib/Api/LearningModelSession.h:93]: Why was this formatting changed? Is it automatically formatted from a cmake format file? It looks kind of odd and overly verbose.

File: winml/lib/Api/LearningModelSessionOptions.cpp [hunk 7]
```diff
@@ -52,4 +52,14 @@ STDMETHODIMP LearningModelSessionOptions::SetIntraOpThreadSpinning(boolean allow
   telemetry_helper.SetIntraOpThreadSpinning(allowSpinning != 0);
   return S_OK;
 }
+
+const std::vector<winrt::hstring>& LearningModelSessionOptions::GetCustomOpLibraryPaths() noexcept {
```
Comment (PatriceVignola) [inline winml/lib/Api/LearningModelSessionOptions.cpp:56]: can we return a span of `const winrt::hstring` instead here?

File: winml/lib/Api/LearningModelSessionOptions.h [hunk 8]
```diff
@@ -31,6 +31,14 @@ struct LearningModelSessionOptions : LearningModelSessionOptionsT<LearningModelS
 
   bool GetIntraOpThreadSpinning();
 
+  const std::vector<winrt::hstring>&
```
Comment (PatriceVignola) [inline winml/lib/Api/LearningModelSessionOptions.h:34]: Same thing here, return a span if possible (I'm not sure if this module easily has access to the GSL).

File: winml/api/Microsoft.AI.MachineLearning.Experimental.idl (hunk 1 above)
Comment (smk2007) [reply winml/api/Microsoft.AI.MachineLearning.Experimental.idl:35]: fixed

File: winml/lib/Api.Ort/OnnxruntimeEngine.cpp (hunk 3 above)
Comment (smk2007) [reply winml/lib/Api.Ort/OnnxruntimeEngine.cpp:433]: used reinterpret_cast

File: winml/api/Microsoft.AI.MachineLearning.Experimental.idl (hunk 2 above)
Comment (smk2007) [reply winml/api/Microsoft.AI.MachineLearning.Experimental.idl:148]: removed double space

File: winml/lib/Api/LearningModelSession.cpp (hunk 5 above)
Comment (smk2007) [reply winml/lib/Api/LearningModelSession.cpp:142]: moved brace to same line

File: winml/lib/Api/LearningModelSessionOptions.h (hunk 8 above)
Comment (smk2007) [reply winml/lib/Api/LearningModelSessionOptions.h:34]: done

File: winml/lib/Api/LearningModelSessionOptions.cpp (hunk 7 above)
Comment (smk2007) [reply winml/lib/Api/LearningModelSessionOptions.cpp:56]: doone

File: winml/lib/Api/LearningModelSession.h (hunk 6 above)
Comment (smk2007) [reply winml/lib/Api/LearningModelSession.h:93]: used in other places for consistency

File: winml/lib/Api.Ort/OnnxruntimeEngineBuilder.cpp (hunk 4 above)
Comment (smk2007) [reply winml/lib/Api.Ort/OnnxruntimeEngineBuilder.cpp:61]: const-ed

File: winml/lib/Api.Ort/OnnxruntimeEngine.cpp [hunk 9]
```diff
@@ -428,6 +428,17 @@ HRESULT OnnxruntimeEngine::RuntimeClassInitialize(OnnxruntimeEngineFactory* engi
...
+  }
+}
+
+HRESULT OnnxruntimeEngine::RegisterCustomOpLibraryHandles(const std::vector<void*>& handles) {
```
[the remark this answers is not in the document]
Comment (smk2007) [reply winml/lib/Api.Ort/OnnxruntimeEngine.cpp:437]: left as reference to vector as the caller will need push/pull on the container. So avoiding unnecessary span construction.

[the remark this answers is not in the document]
Comment (PatriceVignola) [reply winml/lib/Api.Ort/OnnxruntimeEngine.cpp:437]: But the whole point of spans is that it's super lightweight to copy (even more than a reference), and callers don't need to explicitly construct the span (they can just pass their usual vector, but also any container they like).

[the remark this answers is not in the document]
Comment (smk2007) [reply winml/lib/Api.Ort/OnnxruntimeEngine.cpp:437]: But it cant be faster than just passing by reference.
In any case i dont think it affects much, changes to gsl::span

**Review by PatriceVignola — APPROVED** [review]

**Pull request merged**